src/tools/rbd_mirror/image_map/types.h

```diff
@@ -0,0 +1,161 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_RBD_MIRROR_IMAGE_MAP_TYPES_H
+#define CEPH_RBD_MIRROR_IMAGE_MAP_TYPES_H
+
+#include <set>
+#include <map>
+#include <string>
+#include <memory>
+
+namespace rbd {
+namespace mirror {
+namespace image_map {
+
+enum MapShuffleType {
+  SHUFFLE_TYPE_INSTANCES_ADDED = 0,   // rebalance due to addition of instances
+  SHUFFLE_TYPE_INSTANCES_REMOVED,     // rebalance due to removal of instances
+  SHUFFLE_TYPE_IMAGES_ADDED_REMOVED,  // rebalance due to adding/removing images
+};
+
+struct MappedImage {
+  std::string id;
```

Comment:
What is this id used for? It's not obvious w/o tracing back through the code. As I mentioned in the (obsolete) ```ImageSpec``` class, I think you really need to track a set of peer UUIDs (black for local pool, populated UUID for remote peer pools). Once you drop to zero UUIDs, you know you can unmap the image from an instance.